packages/@headlessui-react/src/components/combobox/combobox.tsx

```diff
@@ -1707,23 +1708,39 @@ function OptionsFn<TTag extends ElementType = typeof DEFAULT_OPTIONS_TAG>(
     onMouseDown: handleMouseDown,
   })
 
+  // We should freeze when the `visible` state is true and if the `visible`
+  // state and the `comboboxState` are not in sync. This means that a transition
+  // is happening and the component is still visible (for the transition effect)
+  // but closed from a functionality perspective.
+  let shouldFreeze = visible && data.comboboxState === ComboboxState.Closed
+
+  let options = useFrozenData(shouldFreeze, data.virtual?.options)
+
   // Map the children in a scrollable container when virtualization is enabled
-  if (data.virtual && visible) {
+  if (data.virtual) {
```

Comment:
This was actually switching between virtual and non-virtual mode if the `visible` state changes 😬